posts/using-ansible-and-ecs-fargate-for-an-easy-client-handoff.md

```diff
@@ -0,0 +1,79 @@
+
+## Managing automation extremes--flexible vs simple ##
+
+```
+Automate everything.
+
+			-- Every DevOps talk, blog, and whitepaper.
+```
+
+There's a lot of pressure out there to automate, scale, and be infinitely flexible. There's also a lot of pressure to ship.  Don't get tangled in trying to automate everythign.  If you don't have some frameworks in play, don't expect to have much automation bundled into your project.  It takes time.  Balance what is flexible vs what just needs to work, while insulating yourself from the long-term impact of any [tempermanent](https://www.urbandictionary.com/define.php?term=Tempermanent) solutions.   
+
+## Decisions that worked for us ##
+
+### Discrete front-end and back-end microservices with containers ###
+This is kind of microservices 101, but go ahead and separate front-end code early.   They will deploy and scale differently.  Segmenting also makes it easy to get truthful metrics about network traffic and system load.
+
+### Simple automated builds and redeployment with Gitlab-CI ###
+Gitlab-CI's YAML job descriptions are really quick to implement and modify.  Our job would build, test, push to repo, and call for the service to redeploy in 50 lightweight lines of YAML.  The client was able to sign up for the base-level hosted Gitlab offering, and their engineer was able to drop in the repo and CI integration.
+
+### AWS Fargate to minimize container management overhead ###
+AWS Fargate has its place.  IMHO that's for small green field projects, where there's a very basic need to run containers without persistent storage, and no other infrastructure to integrate.  Our client's project met this criteria with flying colors. Ansible 2.7 has Fargate support, which further sweetened the proposition.
+
+### Application routing delegated to ALB Rules ###
+Rather than the front end forward API requests to the backend nodes, we let ALB rules route traffic to the appropriate front or backend target group based on URL.  Service discovery was not needed since the balancer took care of routing to the different services.   Deployments instantly become transparent because ALB's healthchecks will respond to a node's availability.
+
+### Database connection information stored in AWS Systems Manger Parameter Store ###
+Since we were using Spring Boot, we had a lot of configuration methods to choose from.   Due to the small scale of the application, we opted for 12-factor style config by setting environment variables for the JDBC path and credentials.  It did not require any code changes to use.  AWS Systems Manager Parameter Store was a simple solution that could easily be extended for more configurations.  To inject the configuration, a function was added to the docker entrypoint shell script to extract the values from the parameter store and inject into the environment.
```

Comment:
```suggestion
Since we were using Spring Boot, we had a lot of configuration methods to choose from.   Due to the small scale of the application, we opted for 12-factor style config by setting environment variables for the JDBC path and credentials.  It did not require any code changes to use.  AWS Systems Manager Parameter Store was a simple solution that could easily be extended for more configurations.  To inject the configuration, a function was added to the Docker entrypoint shell script to extract the values from the parameter store and inject into the environment.
```